File: src/warmup.ts
import OpenAI from 'openai';
import type { ChatCompletionTool } from 'openai/resources/chat/completions.js';

export interface WarmupArgs {
  client: OpenAI;
  model: string;
  systemPrompt: string;
  tools: ChatCompletionTool[];
}

// Send a minimal request with the same system prompt + tools we'll use
// for real turns. llama.cpp/Ollama cache the prompt-eval result for
// matching prefixes, so the first real user prompt only needs to evaluate
// the new user tokens — turns ~12s of cold prompt-eval into ~2s.
export async function warmup(args: WarmupArgs): Promise<{ ms: number }> {
  const start = Date.now();
  await args.client.chat.completions.create({
    model: args.model,
    messages: [
      { role: 'system', content: args.systemPrompt },
      { role: 'user', content: 'ok' },
    ],
    tools: args.tools.length ? args.tools : undefined,
    max_tokens: 1,
    stream: false,
  });
  return { ms: Date.now() - start };
}

File: src/warmup.rs
use anyhow::Result;
use std::time::Instant;

use crate::backends::BackendDescriptor;
use crate::model_system::EffortLevel;
use crate::openai::{chat_oneshot, ChatMessage, ChatRequest, ToolDef};

pub async fn warmup(
    http: &reqwest::Client,
    backend: &BackendDescriptor,
    model: &str,
    effort: Option<EffortLevel>,
    system_prompt: &str,
    tools: &[ToolDef],
) -> Result<u128> {
    let start = Instant::now();
    let messages = vec![
        ChatMessage::System {
            content: system_prompt.to_string(),
        },
        ChatMessage::User {
            content: "ok".into(),
        },
    ];
    let cache_key = crate::openai::session_cache_key(backend, model, &messages);
    let req = ChatRequest {
        model,
        messages: &messages,
        tools: if tools.is_empty() { None } else { Some(tools) },
        stream: false,
        stream_options: None,
        max_tokens: Some(1),
        // Use the same routing key as the real turn; otherwise a successful
        // warmup may populate a different OpenAI cache shard.
        prompt_cache_key: cache_key.as_deref(),
        effort,
    };
    chat_oneshot(http, backend, &req).await?;
    Ok(start.elapsed().as_millis())
}
